internal/upload/oci/upload.go

```diff
@@ -0,0 +1,234 @@
+		CreateImageDetails: core.CreateImageDetails{
+			DisplayName:   common.String(imageName),
+			CompartmentId: common.String(compartmentID),
+			FreeformTags: map[string]string{
+				"Uploaded-By": "osbuild-composer",
+				"Name":        imageName,
+			},
+			ImageSourceDetails: core.ImageSourceViaObjectStorageTupleDetails{
+				BucketName:      common.String(bucketName),
+				NamespaceName:   common.String(namespace),
+				ObjectName:      common.String(objectName),
+				SourceImageType: core.ImageSourceDetailsSourceImageTypeQcow2,
+			},
+		},
+	}
+
+	createImageResponse, err := c.computeClient.CreateImage(context.Background(), request)
+	if err != nil {
+		return "", fmt.Errorf("failed to create an image from storage object: %w", err)
+	}
+
+	log.Printf("waiting for the work request to complete, this may take a while. Work request ID: %s\n", *createImageResponse.OpcWorkRequestId)
+	for {
+		r, err := c.workRequestsClient.GetWorkRequest(context.Background(), workrequests.GetWorkRequestRequest{
+			WorkRequestId: createImageResponse.OpcWorkRequestId,
+			OpcRequestId:  createImageResponse.OpcRequestId,
+		})
+		if err != nil {
+			return "", fmt.Errorf("failed to fetch the work request for creating the image: %w", err)
+		}
+		switch r.Status {
+		case workrequests.WorkRequestStatusSucceeded:
+			return *createImageResponse.Id, nil
+		case workrequests.WorkRequestStatusCanceled, workrequests.WorkRequestStatusFailed:
+			return "", fmt.Errorf("the work request for creating an image is in status %s", r.Status)
+		}
+		log.Printf("work request - status: %s, percent completed: %f\n", r.Status, *r.PercentComplete)
+		time.Sleep(10 * time.Second)
+	}
+	return *createImageResponse.Id, nil
+}
+
+type ClientParams struct {
+	User        string
+	Region      string
+	Tenancy     string
+	PrivateKey  string
+	Fingerprint string
+}
+
+type ociClient struct {
+	storageClient      objectstorage.ObjectStorageClient
+	identityClient     identity.IdentityClient
+	computeClient      core.ComputeClient
+	workRequestsClient workrequests.WorkRequestClient
+}
+
+func (c Client) GetBucket(bucketName, namespace string) (objectstorage.Bucket, error) {
+	res, err := c.storageClient.GetBucket(context.Background(), objectstorage.GetBucketRequest{
+		NamespaceName:   common.String(namespace),
+		BucketName:      common.String(bucketName),
+		RequestMetadata: common.RequestMetadata{},
+	})
+	if err != nil {
+		return objectstorage.Bucket{}, err
+	}
+	return res.Bucket, nil
+}
+
+// GrantImageAccess grants the user the needed permissions to launch an instance
+// from the imageID in the designated compartment
+func (c Client) GrantImageAccess(imageID string, user string, compartmentID string) error {
+	// TBD
+	// should we support granting access to a group? oci have an api to grant
+	// access to groups or any-user (all users in tenancy) rather than specific user.
+	// also, not all regions support identity domains, i.e. they'll need different method to
+	// allow access.
```

Comment:
I don't know what the best approach is here.  We'd have to look into what's generally considered a sane default for OCI, I guess.